firedrake/preconditioners/asm.py

```diff
@@ -110,6 +112,12 @@ def view(self, pc, viewer=None):
         self.asmpc.view(viewer=viewer)
 
     def update(self, pc):
+        # This is required to update an inplace ILU/ICC factorization
+        try:
+            for sub in self.asmpc.getASMSubKSP():
+                sub.getOperators()[0].setUnfactored()
+        except PETSc.Error:
```

Comment:
This seems like a vague `except`. Also what if `self.asmpc.getASMSubKSP()` has 2 entries and the first errors but the second would not?

Reply:
I cannot reproduce the conditions to get that PETSc exception, it was a while ago, it could be that I fixed this within PETSc. But the important thing is that setUnfactored must be called. Ideally it should happen in PETSc, but not sure where. Another important detail is that we don't need to call `self.asmpc.setUp()` here as PETSc should know that this should happen if the matrix is reassambled.